service/controller/app/v1/resource/chartoperator/resource.go

```diff
@@ -0,0 +1,153 @@
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/helmclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/spf13/afero"
+	"gopkg.in/yaml.v2"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/helm/pkg/helm"
+
+	"github.com/giantswarm/app-operator/service/controller/app/v1/key"
+)
+
+const (
+	Name = "chartoperatorv1"
+
+	defaultClusterDNSIP    = "172.31.0.10"
+	chartOperatorNamespace = "giantswarm"
+	chartOperatorRelease   = "chart-operator"
+	chartOperatorVersion   = "0.9.0"
+)
+
+// Config represents the configuration used to create a new clients resource.
+type Config struct {
+	// Dependencies.
+	Fs        afero.Fs
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	// Settings.
+	RegistryDomain string
+}
+
+type Resource struct {
+	// Dependencies.
+	fs        afero.Fs
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+
+	// Settings.
+	registryDomain string
+}
+
+// New creates a new configured chartoperator resource.
+func New(config Config) (*Resource, error) {
+	if config.Fs == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Fs must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.RegistryDomain == "" {
+		config.RegistryDomain = "quay.io"
+	}
+
+	r := &Resource{
+		// Dependencies.
+		fs:        config.Fs,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		// Settings.
+		registryDomain: config.RegistryDomain,
+	}
+
+	return r, nil
+}
+
+func (r Resource) Name() string {
+	return Name
+}
+
+func (r Resource) installChartOperator(ctx context.Context, cr v1alpha1.App, helmClient helmclient.Interface) error {
+	var err error
+
+	var tarballPath string
+	{
+		tarballURL := fmt.Sprintf("https://giantswarm.github.io/giantswarm-catalog/chart-operator-%s.tgz", chartOperatorVersion)
+		tarballPath, err = helmClient.PullChartTarball(ctx, tarballURL)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		defer func() {
+			err := r.fs.Remove(tarballPath)
+			if err != nil {
+				r.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("deletion of %#q failed", tarballPath), "stack", fmt.Sprintf("%#v", err))
+			}
+		}()
+	}
+
+	var clusterDNSIP string
+	{
+		name := key.ClusterValuesConfigMapName(cr)
+		cm, err := r.k8sClient.CoreV1().ConfigMaps(cr.Namespace).Get(name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("no cluster-value %#q in control plane, operator will use default clusterDNSIP value", name))
+			clusterDNSIP = defaultClusterDNSIP
```

Comment:
Why should that actually ever happen? And in case it ever happens, why can we set the value statically? If I am not mistaken we derive the DNS IP from the Cluster IP range and this is configured via the installations repo. So I assume it can be different. I would either drop the defaulting here as I do not see why we need it, but if I am missing something the defaulting should not be done based on some static value. Or as a third option, what we do in the `cluster-operator` is not necessary but then we should clean up the mess there.

Reply:
Yes like I replied before this is for bootstrapping CoreDNS. That is only needed when chart-operator is running in a TC.

If chart-operator is running locally in kind or minikube then there is already a CoreDNS. So yes I think we should drop this.

Reply:
So I found out we don't need to have default value as separate, we keep this value in chart values.